Add reprepro to debarchive migration instructions (#227)
canonical/landscape-documentation: docs/how-to-guides/upgrade/migrate-repository-mirrors-to-debarchive.md

Reviewer: I'm noticing that the examples don't have all the placeholders called out - only some are marked placeholders, and some use examples directly provided at the beginning of the step.

Can you go through the examples and make sure that's clear? They could just be resolved by clarifying it's an example call. Like instead of "create the mirror", change it to "create the mirror. Example call:" (or similar)

I like providing the examples for syntax rather than just placeholders everywhere, but how it is now, a user could copy it, change the placeholders, and not notice there's other "example" data inside it (components, filters, etc.)

Author: Yes, I'll do this, but I'll leave it to last, as it's a bit more involving than the other changes you've commented on.

Author: I've added the "example call"-style wording. I'm not sure it's completely clarifies it, but I think it helps a bit. FWIW, Joey didn't seem to have trouble identifying these as examples that should be edited.

Author: But maybe he's just really smart.

Reviewer: Joey IS really smart :D but also I don't think most people would have issues, but some might and I want to reduce the friction if people are just copying codeblocks. It's not a big annoyance if you copy something and it fails, then you fix it, but still those annoyances add up ..... 😇

## Identify your existing repositories

Before migrating, inventory your existing reprepro-managed repositories. On the Landscape Server, the reprepro repository data is stored under `/var/lib/landscape/landscape-repository/standalone/`.

List all configured distributions and their pockets. Each distribution has its own configuration directory:

```bash
cat /var/lib/landscape/landscape-repository/standalone/<DISTRIBUTION>/conf/distributions
```

where `<DISTRIBUTION>` is the name of the distribution you created (e.g., `ubuntu`, `ubuntu-staging`).

This file contains stanzas like:

```text
Codename: noble-release
Components: main restricted universe multiverse
Architectures: amd64
SignWith: <KEY_FINGERPRINT>
```

```text
Codename: noble-updates
Components: main restricted universe multiverse
Architectures: amd64
SignWith: <KEY_FINGERPRINT>
Update: noble-updates
```

```text
Codename: noble-staging
Components: main
Architectures: amd64
SignWith: <KEY_FINGERPRINT>
```

The `Update:` field indicates a sync (mirror) pocket. Stanzas with a `Pull:` field are pull pockets. Stanzas without either are upload pockets.

To see what packages are currently in a pocket:

```bash

> reprepro -b /var/lib/landscape/landscape-repository/standalone/<DISTRIBUTION> list noble-release

Reviewer 2: nit: Use `<Codename>` instead of `noble-release`?

  -H "Content-Type: application/json" \
  -d '{
    "displayName": "noble-release-staging",
    "archiveRoot": "http://archive.ubuntu.com/ubuntu",
    "distribution": "noble",
    "architectures": ["amd64"],
    "components": ["main"],
    "filter": "Name (= nginx) | Name (= curl) | Name (= libssl3)",
    "filterWithDeps": true
  }'
```

Set `filterWithDeps` to `true` if you want the filter to also include dependencies of matched packages (recommended for allowlists).

### 4. Sync the filtered mirror

```bash
curl -X POST "$API_BASE/mirrors/<MIRROR_ID>:sync" \
  -H "Authorization: Bearer $JWT" \
  -H "Content-Type: application/json"
```

## Migrate upload pockets

Upload pockets contain packages that were manually uploaded by users. Since there's no upstream source to sync from, you need to import the existing packages into a new local repository.

### 1. Identify the packages

List the packages in the upload pocket:

```bash
reprepro -b /var/lib/landscape/landscape-repository/standalone/<DISTRIBUTION> list noble-staging
```

The pool directory containing the actual `.deb` files is at:

```
/var/lib/landscape/landscape-repository/standalone/<DISTRIBUTION>/pool/
```

> 

Reviewer 2: This is true, but for me it is also listing all the packages from my other pockets. It's probably because I did `noble-backports universe` for all my pockets but raising in case you think there should be a change here.

Reviewer 2: Okay there's a note for this later, but I think the info could be useful here.

### 2. Make packages accessible via URL

The `ImportLocalPackages` API accepts a URL pointing to a `.deb` file. You need to make your existing packages accessible via HTTP or a `file://` URL.

If the Deb Archive service runs on the same machine as your existing repository, you can use `file://` URLs directly:

```bash
# Example: find all .deb files for the staging pocket's component
find /var/lib/landscape/landscape-repository/standalone/<DISTRIBUTION>/pool/main/ -name "*.deb"
```

Alternatively, if the packages are served by the existing Landscape repository web server, they may already be accessible at:

```
https://$LANDSCAPE_FQDN/repository/standalone/<DISTRIBUTION>/pool/

> ```

Reviewer 2: This didn't work for me though it only says `"may"`. I just got a Landscape 404 page.

Reviewer 2: I think this worked when I swapped to `http` instead of `https`. I think because my browser doesn't trust the cert so it may not be an issue for deployments with actual domains.

Author: I think the default apache setup only serves this via HTTP. I will change it.

```{note}
When using `file://` URLs, the path must be accessible to the `landscape-debarchive` snap. Because snap confinement restricts filesystem access, `file://` URLs pointing to arbitrary paths outside the snap's data directory will not work. Packages must instead be served over HTTP — either by the Landscape Server itself, or by a temporary HTTP server.
```

If the Landscape Server is already serving the repository, packages are accessible at:

```
http://$LANDSCAPE_FQDN/repository/standalone/<DISTRIBUTION>/pool/m/my-package/my-package_1.0-1_amd64.deb
```

Alternatively, start a temporary HTTP server:

```bash
python3 -m http.server 8080 --directory /path/to/packages
```

Then reference packages as `http://localhost:8080/my-package_1.0-1_amd64.deb`.

```{important}
The reprepro pool directory is shared across all distributions in the same reprepro tree. Scanning or archiving the entire pool will include packages from sync mirror and pull pockets, not just your upload pocket. Use one of the approaches below to import only the packages that belong to this specific pocket.

> ```

Reviewer 2: I think you can move this note up and link to this section or something similar. Personally I had this exact same question but I only came across this note as I was about the copy over only the packages I cared about.

Author: I'm replicating the note further up and slightly adjusted the wording.

#### Script imports from reprepro list

Use `reprepro list` to enumerate only the packages belonging to this pocket, then locate each `.deb` file and import it via HTTP. Set `REPO_HTTP` to the base URL from which the repository is served:

```bash

> #!/bin/bash

Reviewer 2: Script seems to work. We discussed live and some basic error handling would be nice to have.

Author: I'm adding `set -euo pipefail` to the script, which will surface errors better.

Would be nice to have an independent test of it 😉 .

Reviewer 2: Yup the script still works.

  VERSION=$(echo "$line" | awk '{print $3}')

  DEB_PATH=$(find "$DIST_BASE/pool/" \
    \( -name "${PACKAGE}_${VERSION}_*.deb" \
    -o -name "${PACKAGE}_${VERSION//:/%3a}_*.deb" \) | head -1)

  if [ -z "$DEB_PATH" ]; then
    echo "Warning: could not find .deb for $PACKAGE $VERSION"
    continue
  fi

  # Convert local path to HTTP URL
  RELATIVE_PATH="${DEB_PATH#$DIST_BASE/}"
  URL="$REPO_HTTP/$RELATIVE_PATH"

  echo "Importing: $URL"
  RESPONSE=$(curl -s -X POST "$API_BASE/locals/$LOCAL_ID:importPackages" \
    -H "Authorization: Bearer $JWT" \
    -H "Content-Type: application/json" \
    -d "{\"url\": \"$URL\"}")

  OP_NAME=$(echo "$RESPONSE" | jq -r '.name')

  while true; do
    STATUS=$(curl -s -X GET "$API_BASE/$OP_NAME" \
      -H "Authorization: Bearer $JWT" \
      -H "Content-Type: application/json")
    if [ "$(echo "$STATUS" | jq -r '.done')" = "true" ]; then
      break
    fi
    sleep 2
  done
done
```

### 5. Verify imported packages

Once all imports have completed, confirm the packages are present in the local repository:

```bash
curl -X GET "$API_BASE/locals/<LOCAL_ID>/packages" \
  -H "Authorization: Bearer $JWT"

> ```

Reviewer 2: This is really nice, I think it would be useful to have for the other sections.

## Preserve the exact state of a sync mirror

If you need to preserve the precise package versions currently in a sync mirror, rather than re-syncing from upstream (which may have newer versions), treat it as an upload pocket and import the packages into a local repository. Use the steps above for migrating upload pockets, but use the sync mirror pocket instead.

> ## Publish the migrated repositories

Author: We discussed doing these steps in the Landscape UI, after upgrading to 26.04. I think I'll add the upgrade instructions here and have these reference using the UI.

---
myst:
  html_meta:
    description: "Migrate repository mirrors from legacy reprepro-managed repositories (Landscape 25.10 and earlier) to the new debarchive service introduced in Landscape 26.04 LTS."
---

(how-to-migrate-repository-mirrors-to-debarchive)=
# How to migrate repository mirrors to Deb Archive

This guide describes how to migrate your existing reprepro-managed repository mirrors from Landscape 25.10 (and earlier) to the new Deb Archive service introduced in Landscape 26.04 LTS.

```{note}
This guide assumes you have already:

> - Upgraded to Landscape 26.04 LTS ({ref}`how-to-upgrade-to-26-04-lts`)

Reviewer 2: Do we upgrade before starting this guide? It seems like we need to refer back to the reprepro backed deployment to configure Deb Archive, and wouldn't that information be lost if we upgraded? But we need to upgrade to set up  Deb Archive, so how did you intend for the user to have reprepro config available while configuring Deb Archive?

Author: Ah right this is a good point.

The upgrade is blocked anyways if the old distribution records still exist in the database. I think a better workflow would be to:

1. install and configure `landscape-debarchive`
2. perform the migration steps
3. upgrade landscape-server

Reviewer 2: Ah but we can't configure `landscape-debarchive` before upgrading. All services crash if we add debarchive stuff to the `service.conf`. 

```
Extra inputs are not permitted [type=extra_forbidden, input_value='landscape-standalone-debarchive', input_type=str]
May 27 20:31:07 landscape-test log-system-event[76621]:     For further information visit https://errors.pydantic.dev/2.4/v/extra_forbidden
```

I am going to snapshot my 25.10 installation with all the reprepro stuff and take notes of everything needed in the guide and see if that works. Let me know if I'm missing something though.

Author: #226 - uhuh don't set it in the service.conf.

Reviewer 2: Approved https://github.com/canonical/landscape-documentation/pull/226, as we discussed it might make the most sense to move the step about upgrading to  after `Migrate Upload Pockets`

```bash
# Example: find all .deb files for the staging pocket's component
find /var/lib/landscape/landscape-repository/standalone/<DISTRIBUTION>/pool/main/ -name "*.deb"
```

Alternatively, if the packages are served by the existing Landscape repository web server, they may already be accessible at:

```
https://$LANDSCAPE_FQDN/repository/standalone/<DISTRIBUTION>/pool/
```

### 3. Create a local repository

```bash
curl -X POST "$API_BASE/locals" \
  -H "Authorization: Bearer $JWT" \
  -H "Content-Type: application/json" \
  -d '{
    "displayName": "noble-staging",
    "defaultDistribution": "noble",
    "defaultComponent": "main"
  }'
```

The response includes a `localId`.

### 4. Import packages

Import a single package into the local repository:

```bash
curl -X POST "$API_BASE/locals/<LOCAL_ID>:importPackages" \
  -H "Authorization: Bearer $JWT" \
  -H "Content-Type: application/json" \
  -d '{"url": "file:///path/to/my-package_1.0-1_amd64.deb"}'
```

This is a long-running operation. Poll the returned operation for completion before importing the next package.

```{note}

> When using `file://` URLs, the path must be accessible to the `landscape-debarchive` snap. Because snap confinement restricts filesystem access, `file://` URLs pointing to arbitrary paths outside the snap's data directory will not work. Packages must instead be served over HTTP — either by the Landscape Server itself, or by a temporary HTTP server.

Reviewer 2: You're missing the closing ``` for the note

If the Landscape Server is already serving the repository, packages are accessible at:

```
http://$LANDSCAPE_FQDN/repository/standalone/<DISTRIBUTION>/pool/m/my-package/my-package_1.0-1_amd64.deb

> ```

Author: This has been fixed. It should only be HTTP.

Alternatively, start a temporary HTTP server:

```bash
python3 -m http.server 8080 --directory /path/to/packages
```

Then reference packages as `http://localhost:8080/my-package_1.0-1_amd64.deb`.

> ```

Reviewer 2: ```suggestion
```

Author: So that's where those got to…

---
myst:
  html_meta:
    description: "Migrate repository mirrors from legacy reprepro-managed repositories (Landscape 25.10 and earlier) to the new debarchive service introduced in Landscape 26.04 LTS."
---

(how-to-migrate-repository-mirrors-to-debarchive)=
# How to migrate repository mirrors to Deb Archive

This guide describes how to migrate your existing reprepro-managed repository mirrors from Landscape 25.10 (and earlier) to the new Deb Archive service introduced in Landscape 26.04 LTS.

```{note}

> This guide assumes you have already:

Reviewer: You should take these out of a note and put them as a "Prerequisites" section, like https://documentation.ubuntu.com/landscape/how-to-guides/landscape-installation-and-set-up/juju-ha-installation/#prerequisites

(after "Overview")

- Upgraded to Landscape 26.04 LTS ({ref}`how-to-upgrade-to-26-04-lts`)
- Installed and configured the `landscape-debarchive` snap ({ref}`how-to-debarchive-repository-management`)
```

> ## Overview

Reviewer: It'd be helpful to link to https://documentation.ubuntu.com/landscape/explanation/features/repository-mirroring-2604/ in this section.

I'd suggest putting it at the end of the first paragraph or at the end of the full Overview, something like "For details on how repository mirroring works in 26.04, see XYZ"

In Landscape 25.10 and earlier, repository management was handled by an internal reprepro-based system that managed mirrors through distributions, series, and pockets. In Landscape 26.04 LTS, this system is replaced by the `debarchive` service, which provides a REST API for managing mirrors, local repositories, and publications.

The migration strategy depends on the type of pocket you're migrating:

> | Legacy pocket type | Migration strategy |

Reviewer: ```suggestion
| pre-26.04 pocket type | Migration strategy |
```
Thoughts? I wanted to be clearer about what we consider the "legacy" type

In general, I don't think we should use the term "legacy" throughout this guide since it's not a "legacy" system for the user (i.e., it's their current system until they migrate). Can you change those too to pre-26.04 or "existing system" ?

I'm open to other suggestions, but I want to move away from calling it "legacy", or even try to avoid "reprepro-based" in places where it doesn't make sense to explicitly call out reprepro

Author: Seems alright to me. I'll replace mentions of "legacy" with "pre-26.04" throughout.

|---|---|
| **Sync (mirror) pocket** | Create a new mirror in Deb Archive pointing at the original upstream source and sync it |
| **Pull pocket** | Create a new mirror with a `filter` matching your existing allowlist/blocklist |
| **Upload pocket** | Create a local repository and import packages from the existing reprepro pool |

```{important}
If you need to preserve the **exact state** of a sync mirror (the precise set of package versions currently stored, rather than the latest upstream state), you should treat it as an upload pocket and import its packages into a new local repository instead.
```

> ## Prerequisites

Reviewer: I'd consider "Prerequisites" as something you need to have in place and the instructions aren't within the guide. For example: https://documentation.ubuntu.com/landscape/how-to-guides/landscape-installation-and-set-up/juju-ha-installation/#prerequisites

Can you change this heading? Maybe to "Before you begin" or just "Set environment variables"

Author: I'll go with "Set environment variables"

Set the following environment variables for use throughout this guide:

```bash
export LANDSCAPE_FQDN="landscape.example.com"
export API_BASE="https://$LANDSCAPE_FQDN/debarchive/v1beta1"
export JWT="<your-jwt-token>"
```

To obtain a JWT token, authenticate against the Landscape REST API. See {ref}`reference-rest-api-login` for details.

## Identify your existing repositories

Before migrating, inventory your existing reprepro-managed repositories. On the Landscape Server, the reprepro repository data is stored under `/var/lib/landscape/landscape-repository/standalone/`.

List all configured distributions and their pockets. Each distribution has its own configuration directory:

```bash
cat /var/lib/landscape/landscape-repository/standalone/<DISTRIBUTION>/conf/distributions
```

where `<DISTRIBUTION>` is the name of the distribution you created (e.g., `ubuntu`, `ubuntu-staging`).

> This file contains stanzas like:

Reviewer: ```suggestion
This file contains sections like:
```
Or "entry", "block", or"configuration block"/"distribution block" ?

And can you change the other ones? "section" (or whichever term) is more plain English. Be consistent with the term you use (section/block/ .. )

```text
Codename: noble-release
Components: main restricted universe multiverse
Architectures: amd64
SignWith: <KEY_FINGERPRINT>
```

```text
Codename: noble-updates
Components: main restricted universe multiverse
Architectures: amd64
SignWith: <KEY_FINGERPRINT>
Update: noble-updates
```

```text
Codename: noble-staging
Components: main
Architectures: amd64
SignWith: <KEY_FINGERPRINT>
```

The `Update:` field indicates a sync (mirror) pocket. Stanzas with a `Pull:` field are pull pockets. Stanzas without either are upload pockets.

To see what packages are currently in a pocket:

```bash
reprepro -b /var/lib/landscape/landscape-repository/standalone/<DISTRIBUTION> list <CODENAME>
```

Replace `<DISTRIBUTION>` with the appropriate distribution directory name and `<CODENAME>` with the codename of the pocket (e.g., `noble-updates`).

> 

Reviewer: ```suggestion

Use this information to decide which migration sections to follow. For example, if your deployment only has sync pockets, complete the sync-pocket migration steps and skip the pull-pocket and upload-pocket sections.
```

Right? Reword as needed, but something that ties this section together would be helpful, so it's clear what the user should do with this information.

Author: Yeah. I have re-worded slightly, but not by much.

### 2. Create the mirror

```bash
curl -X POST "$API_BASE/mirrors" \
  -H "Authorization: Bearer $JWT" \
  -H "Content-Type: application/json" \
  -d '{
    "displayName": "<CODENAME>",
    "archiveRoot": "http://archive.ubuntu.com/ubuntu",
    "distribution": "<CODENAME>",
    "architectures": ["amd64"],
    "components": ["main", "restricted", "universe", "multiverse"]
  }'
```

The response includes a `mirrorId` that you'll use in subsequent requests.

### 3. Sync the mirror

Trigger a sync to download packages from the upstream source:

```bash
curl -X POST "$API_BASE/mirrors/<MIRROR_ID>:sync" \
  -H "Authorization: Bearer $JWT" \
  -H "Content-Type: application/json"
```

This returns a long-running operation. Poll for its completion:

```bash
curl -X GET "$API_BASE/operations/<OPERATION_ID>" \
  -H "Authorization: Bearer $JWT" \
  -H "Content-Type: application/json"
```

The operation is complete when `done` is `true`.

### 4. Verify the mirror

Once the sync is complete, confirm the packages are present in the mirror:

```bash
curl -X GET "$API_BASE/mirrors/<MIRROR_ID>/packages" \
  -H "Authorization: Bearer $JWT"
```

### 5. Repeat for each sync pocket

> Repeat steps 2–4 for each sync pocket (`release`, `updates`, `security`, etc.).

Reviewer: Is it correct that it's only steps 1-4 (i.e., they all have the same upstream source) .. ? Presumably no, but I don't actually know.

You could also more broadly say "Repeat the previous steps for each sync pocket"

Author: Yes, technically step 1 would be the same for each pocket from the same distribution, but it's also less confusing to just say "do all the steps" even if one is the same. I'll make that change too.

## Migrate pull pockets

> Pull pockets in the legacy system stage packages from another pocket using allowlist or blocklist filters. In Deb Archive, this is achieved by creating a mirror with a `filter` expression that replicates your allowlist or blocklist.

Reviewer: ```suggestion
Pull pockets in the legacy system stage packages from another pocket using allowlist or blocklist filters. Starting in Landscape 26.04 LTS, this is achieved by creating a mirror with a `filter` expression that replicates your allowlist or blocklist, also referred to as a "filtered mirror".
```
Or "Starting in Landscape 26.04 LTS with Deb Archive" if you think it helps to be more explicit.

(1) The user is still experiencing these through _Landscape_, not directly through deb archive
(2) Keeping the terminology consistent - "filtered mirrors"

I think it's okay to call out Deb Archive sometimes in the guide when you are specifically referring to the actual service (e.g., if you're saying where the Deb Archive service runs), but here you're describing a general Landscape behavior

Author: Makes sense. I'm okay with this change as-is.

### 1. Identify the filter rules

Check the reprepro `pulls` configuration:

```bash
cat /var/lib/landscape/landscape-repository/standalone/<DISTRIBUTION>/conf/pulls
```

Example stanza:

```text
Name: noble-release-staging
From: noble-release
Components: main
Architectures: amd64
FilterList: install noble-release-staging.list
```

> The referenced filter list file (e.g., `noble-release-staging.list`) contains package name patterns, one per line. Examine the filter list:

Reviewer: What are they examining it for? (I'd give a clearer suggestion, but I'm not sure what they're looking for)

Author: I think I can remove this "examine" directive. The example below, and the subsequent translation to the new format makes it pretty clear (imo) what they are looking for. Let me know if you disagree.

```bash
cat /var/lib/landscape/landscape-repository/standalone/<DISTRIBUTION>/conf/noble-release-staging.list
```

Example allowlist content:

```text
nginx install
curl install
libssl3 install
```

> ### 2. Translate the filter to a Deb Archive filter expression

Reviewer: ```suggestion
### 2. Migrate allowlists and blocklists to filtered mirrors
```

In general - I found this section confusing. IMO it seems like they're still considered "allowlists" and "blocklists" in 26.04+. But you're trying to say that allow/block lists are now "filtered mirrors", right?

Author: Yes, that's what I'm trying to say. I like your suggestion here.

> Deb Archive mirrors support package filtering using an aptly-compatible filter syntax. Translate your allowlist into a filter expression:

Reviewer: Is there another way you can describe this without saying "aptly-compatible"? I don't want to expose too much of the backend here if we can avoid it.

But if removing Aptly will just make it sound weird / ambiguous, then leave it worded as is and link to Aptly's docs or main product website

Author: I think I can take it out.

  ```
  !Name (= unwanted-package) , !Name (= another-unwanted)
  ```

### 3. Create the filtered mirror

Create a mirror pointing at the same upstream as the original source pocket, with the filter applied:

```bash
curl -X POST "$API_BASE/mirrors" \
  -H "Authorization: Bearer $JWT" \
  -H "Content-Type: application/json" \
  -d '{
    "displayName": "<CODENAME>",
    "archiveRoot": "http://archive.ubuntu.com/ubuntu",
    "distribution": "<CODENAME>",
    "architectures": ["amd64"],
    "components": ["main"],
    "filter": "Name (= nginx) | Name (= curl) | Name (= libssl3)",
    "filterWithDeps": true
  }'
```

Set `filterWithDeps` to `true` if you want the filter to also include dependencies of matched packages (recommended for allowlists).

### 4. Sync the filtered mirror

```bash
curl -X POST "$API_BASE/mirrors/<MIRROR_ID>:sync" \
  -H "Authorization: Bearer $JWT" \
  -H "Content-Type: application/json"
```

### 5. Verify the filtered mirror

> Once the sync is complete, confirm the filtered packages are present in the mirror:

Reviewer: What are they looking for to confirm? If it's obvious from the output, feel free to ignore this suggestion. That's just always a pet peeve of mine when a guide says to check that it worked with XYZ command, but it's not obvious from the output :)

Author: The call lists the packages in the mirror, and the user should read them to make sure they match what their filters were.

I'll see if I can make it a bit more obvious.

```bash
curl -X GET "$API_BASE/mirrors/<MIRROR_ID>/packages" \
  -H "Authorization: Bearer $JWT"
```

## Migrate upload pockets

Upload pockets contain packages that were manually uploaded by users. Since there's no upstream source to sync from, you need to import the existing packages into a new local repository.

### 1. Identify the packages

List the packages in the upload pocket:

```bash
reprepro -b /var/lib/landscape/landscape-repository/standalone/<DISTRIBUTION> list noble-staging
```

> The pool directory containing the actual `.deb` files is at:

Reviewer: Can you clarify what this step / the pool directory is for? (assuming the user won't know why they're looking at the pool directory)

Author: I'll flesh this out a little bit.

  RELATIVE_PATH="${DEB_PATH#$DIST_BASE/}"
  URL="$REPO_HTTP/$RELATIVE_PATH"

  echo "Importing: $URL"
  RESPONSE=$(curl -s -X POST "$API_BASE/locals/$LOCAL_ID:importPackages" \
    -H "Authorization: Bearer $JWT" \
    -H "Content-Type: application/json" \
    -d "{\"url\": \"$URL\"}")

  OP_NAME=$(echo "$RESPONSE" | jq -r '.name')

  while true; do
    STATUS=$(curl -s -X GET "$API_BASE/$OP_NAME" \
      -H "Authorization: Bearer $JWT" \
      -H "Content-Type: application/json")
    if [ "$(echo "$STATUS" | jq -r '.done')" = "true" ]; then
      break
    fi
    sleep 2
  done
done
```

### 5. Verify imported packages

Once all imports have completed, confirm the packages are present in the local repository:

```bash
curl -X GET "$API_BASE/locals/<LOCAL_ID>/packages" \
  -H "Authorization: Bearer $JWT"
```

## Preserve the exact state of a sync mirror

If you need to preserve the precise package versions currently in a sync mirror, rather than re-syncing from upstream (which may have newer versions), treat it as an upload pocket and import the packages into a local repository. Use the steps above for migrating upload pockets, but use the sync mirror pocket instead.

> ## Upgrade to Landscape 26.04 LTS

Reviewer: I'm not sure what the correct workflow should be, but this contradicts the prereqs listed at the top which says they should have already upgraded to 26.04?

Author: Right. We pivoted on this partway through.

It's much cleaner to upgrade at this point than before doing any of the steps. I'll remove the upgrade from the (new) prereqs section.

Reviewer: Ok, in that case, this should be mentioned in the full upgrade guide.

I think the easiest way to approach this (given what we currently have) is to add a new "If you use repository management" (or similar title) section below the intro / before [Begin your upgrade](https://documentation.ubuntu.com/landscape/how-to-guides/upgrade/upgrade-to-26-04-lts/#begin-your-upgrade) that clarifies if you're using Landscape to manage repositories, then you should start in the Migrate guide

The alternative would be to have two full upgrade guides - one for with repository management and one for without repository management that have duplicated content. I think the most practical approach given what we currently have is to do my first suggestion (add a section to the existing upgrade guide that links to this migrate guide), but still pointing this out if you disagree with the first option.

Once you have confirmed that all packages are present in the new Deb Archive service, clean up the legacy reprepro Distribution records and then upgrade.

### 1. Delete existing reprepro Distribution records

> In the Landscape web UI, navigate to **Repositories**. Each distribution (for example, `ubuntu` or `ubuntu-staging`) has a **Delete** button. Delete each distribution.

Reviewer: all "web UI" should be "web portal"

### 1. Create a publication target

In the Landscape web UI, navigate to **Repositories > Publication Targets** and create a new publication target. For a filesystem target that serves repositories from the Landscape Server itself, set the path to the directory where published repositories should be written.

### 2. Create a publication

Navigate to **Repositories > Publications** and create a new publication. Select the migrated mirror or local repository as the source, choose the publication target you created, and configure the distribution, architectures, and GPG signing key.

### 3. Publish

Trigger the publication. Landscape will generate the repository metadata and make the packages available at the publication target path.

## Update client machines

After publishing, update the repository profiles on your managed machines to point to the new Deb Archive publication URLs instead of the legacy repository paths.

The legacy repository was served at:

```
deb https://$LANDSCAPE_FQDN/repository/standalone/ubuntu <codename>-<pocket> <components>
```

> The new Deb Archive filesystem publications can be served from the configured published root path using the web server of your choice. Consult your publication target configuration for the exact URL.

Reviewer: "the web server of your choice" should change, but I'm not sure what to. 

At this point, did they already choose the web server they're using when they did the upgrade? If yes:

"the web server you set up before" (or something like that - reference it's something they already did)

If they haven't, then

"a web server, such as nginx or Apache" (or something like that - give them a nudge)

Author: I like the "such as" part. We may end up adding a small web server to the deb archive snap, based on Field DC feedback, so I'll need to come back and change this a bit more, but for now I think it's fine.

## Overview

In Landscape 25.10 and earlier, repository management was handled by an internal reprepro-based system that managed mirrors through distributions, series, and pockets. In Landscape 26.04 LTS, this system is replaced by the `debarchive` service, which provides a REST API for managing mirrors, local repositories, and publications.

The migration strategy depends on the type of pocket you're migrating:

| Pre-26.04 pocket type | Migration strategy |
|---|---|
| **Sync (mirror) pocket** | Create a new mirror in Deb Archive pointing at the original upstream source and sync it |
| **Pull pocket** | Create a new mirror with a `filter` matching your existing allowlist/blocklist |
| **Upload pocket** | Create a local repository and import packages from the existing reprepro pool |

```{important}
If you need to preserve the **exact state** of a sync mirror (the precise set of package versions currently stored, rather than the latest upstream state), you should treat it as an upload pocket and import its packages into a new local repository instead.
```

> For details on how repository mirroring works in 26.04, see {ref}` "Repository mirroring (26.04 LTS and later)"<explanation-repo-mirroring-2604>`.

Reviewer: They syntax is off - can you remove the quotation marks? 

<img width="793" height="82" alt="Image" src="https://github.com/user-attachments/assets/7b1de04e-a644-45bb-a1f5-450670221abd" />

Author: Oops, yeah.

## Prerequisites

This guide assumes you have already:

- Installed and configured the `landscape-debarchive` snap ({ref}`how-to-debarchive-repository-management`)

## Set environment variables

> Set the following environment variables for use throughout this guide:

Reviewer: ```suggestion
Set the following environment variables for use throughout this guide, using the FQDN of your Landscape deployment:
```
or something like that - just wanted to call it out so they don't accidentally take example.com

Author: Sounds good. I'll also mention that it's a JWT for the REST API.

Author: Ugh also realized that "JWT token" is an "ATM Machine" situation, but also just "JWT" sounds awkward.
